cns/restserver/ipam.go

```diff
@@ -478,101 +639,228 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+		//nolint:goerr113 // return error
+		return podIPInfo, fmt.Errorf("not enough desired IPs found in pool")
+	}
+
+	failedToAssignIP := false
+	// assigns all IPs that were found as available to the pod
+	for i := range ipConfigsToAssign {
+		if err := service.assignIPConfig(ipConfigsToAssign[i], podInfo); err != nil {
+			logger.Errorf(err.Error())
+			failedToAssignIP = true
+			break
+		}
+		if err := service.populateIPConfigInfoUntransacted(ipConfigsToAssign[i], &podIPInfo[numIPConfigsAssigned]); err != nil {
+			logger.Errorf(err.Error())
+			failedToAssignIP = true
+			break
+		}
+		// adds to the newly assigned IP to the counter
+		numIPConfigsAssigned++
+	}
+
+	// if we were able to get at least one IP but not all of the desired IPs
+	if failedToAssignIP {
+		logger.Printf("[AssignDesiredIPConfigs] Failed to retrieve all desired IPs. Releasing all IPs that were found")
+		for i := range ipConfigsToAssign {
+			_, err := service.unassignIPConfig(ipConfigsToAssign[i], podInfo)
+			if err != nil {
+				logger.Errorf("[AssignDesiredIPConfigs] failed to mark IPConfig [%+v] back to Available. err: %v", ipConfigsToAssign[i], err)
+			}
+		}
+		//nolint:goerr113 // return error
+		return podIPInfo, fmt.Errorf("not all requested ips %v were found/available in the pool", desiredIPAddresses)
+	}
+
+	logger.Printf("[AssignDesiredIPConfigs] Successfully assigned all desired IPs for pod %+v", podInfo)
+	return podIPInfo, nil
 }
 
-func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) (cns.PodIpInfo, error) {
+// Assigns an available IP from each NC on the NNC. If there is one NC then we expect to only have one IP return
+// In the case of dualstack we would expect to have one IPv6 from one NC and one IPv4 from a second NC
+func (service *HTTPRestService) AssignAvailableIPConfigs(podInfo cns.PodInfo) ([]cns.PodIpInfo, error) {
 	service.Lock()
 	defer service.Unlock()
-
+	// Sets the number of IPs needed equal to the number of NCs so that we can get one IP per NC
+	numIPsNeeded := len(service.state.ContainerStatus)
+	// Creates a slice of PodIpInfo with the size as number of NCs to hold the result for assigned IP configs
+	podIPInfo := make([]cns.PodIpInfo, numIPsNeeded)
+	// This map is used to store whether or not we have found an available IP from an NC when looping through the pool
+	ipsToAssign := make(map[string]cns.IPConfigurationStatus)
+
+	// Searches for available IPs in the pool
 	for _, ipState := range service.PodIPConfigState {
-		if ipState.GetState() == types.Available {
-			if err := service.assignIPConfig(ipState, podInfo); err != nil {
-				return cns.PodIpInfo{}, err
-			}
+		// check if an IP from this NC is already set side for assignment.
+		_, ncAlreadyMarkedForAssignment := ipsToAssign[ipState.NCID]
+		// Checks if we haven't already found an IP from that NC and checks if the current IP is available
+		if ncAlreadyMarkedForAssignment || ipState.GetState() != types.Available {
+			continue
+		}
```

Comment:
combining the logic like this does not save any resoures and just makes it harder to read.
```suggestion
		// check if an IP from this NC is already set side for assignment.
		if _, ncAlreadyMarkedForAssignment := ipsToAssign[ipState.NCID]; ncAlreadyMarkedForAssignment {
			continue
		}
		// Checks if the current IP is available
		if ipState.GetState() != types.Available {
			continue
		}
```